Feature/json payload support mutating policy

yrsuthari opened the pull request:
# Add JSON Payload Support to MutatingPolicy

## Description

This PR implements support for JSON (non-Kubernetes) payloads in MutatingPolicy, addressing issue #13751. The enhancement enables MutatingPolicy to process arbitrary JSON data, not just Kubernetes resources, significantly expanding its use cases.

## Changes Made

### Core Implementation
- **Modified `Handle()` method** in `pkg/cel/policies/mpol/engine/engine.go` to detect and process `JsonPayload` field from `EngineRequest`
- **Updated `MatchedMutateExistingPolicies()` method** to support JSON payload processing
- **Added automatic payload type detection** - engine now seamlessly handles both Kubernetes admission requests and JSON payloads
- **Created minimal admission attributes** for JSON payload processing to maintain compatibility with existing policy evaluation pipeline

### Key Features
- **Backward Compatible**: No changes required to existing policies or workflows
- **Automatic Detection**: Engine automatically detects payload type (K8s vs JSON)
- **Seamless Processing**: JSON payloads processed through same mutation pipeline as K8s resources
- **Policy Reuse**: Existing MutatingPolicies can be applied to JSON payloads

### Testing
- **Added comprehensive test suite** with multiple scenarios:
  - Basic JSON payload handling
  - JSON payload processing without policies
  - MatchedMutateExistingPolicies with JSON payloads
- **Verified backward compatibility** - all existing tests continue to pass
- **Added integration examples** demonstrating real-world usage

### Documentation
- **Created detailed README** with usage examples and implementation details
- **Provided integration test examples** showing how to use the feature
- **Documented key use cases** and benefits

## Use Cases Enabled

This enhancement opens up new possibilities for MutatingPolicy:

1. **Non-Kubernetes JSON Configuration Processing**
   - Apply policies to application configs, API payloads, etc.

2. **API Request/Response Mutation**
   - Validate and mutate REST API payloads

3. **Custom Data Format Processing**
   - Process any JSON data through Kyverno's policy engine

4. **External System Integration**
   - Apply policies to data from external systems producing JSON

## Technical Implementation

The implementation leverages the existing `JsonPayload` field in `engine.EngineRequest` and the `RequestFromJSON()` function. The engine now:

```go
// Detect payload type
if request.JsonPayload != nil {
    // Handle non-K8s JSON payload
    object = *request.JsonPayload
    // Create minimal admission request for policy evaluation
} else {
    // Handle Kubernetes admission request (existing flow)
    object, oldObject, err = admissionutils.ExtractResources(nil, request.Request)
}
```

## Example Usage

```go
// Create a JSON payload (non-Kubernetes)
jsonPayload := &unstructured.Unstructured{
    Object: map[string]interface{}{
        "user": map[string]interface{}{
            "name":  "john",
            "email": "john@example.com",
        },
        "settings": map[string]interface{}{
            "theme": "light",
        },
    },
}

// Create engine request from JSON payload
request := engine.RequestFromJSON(contextProvider, jsonPayload)

// Process with MutatingPolicy engine
response, err := mpolEngine.Handle(context.Background(), request, nil)
```

## Testing Results

```bash
All new tests passing:
- TestHandleJSONPayload/basic_JSON_payload_handling
- TestHandleJSONPayload/JSON_payload_without_policies  
- TestMatchedMutateExistingPoliciesJSONPayload

All existing tests passing (no regressions):
- 20+ existing MutatingPolicy engine tests
- CEL engine compatibility tests
```

## Files Changed

- `pkg/cel/policies/mpol/engine/engine.go` - Core implementation
- `pkg/cel/policies/mpol/engine/engine_test.go` - Test suite
- `examples/mutating-policy-json-payloads/README.md` - Documentation
- `examples/mutating-policy-json-payloads/example_test.go` - Integration examples

## Impact

- **No breaking changes** - fully backward compatible
- **Extends functionality** without affecting existing workflows  
- **Minimal performance overhead** - only processes JSON payloads when provided
- **Maintains existing API** - no changes to public interfaces

## Checklist

- [x] Implementation completed and tested
- [x] All existing tests pass
- [x] New tests added for JSON payload scenarios
- [x] Documentation and examples provided
- [x] Backward compatibility verified
- [x] Code follows project conventions

## Related Issue

Fixes #13751

## Review Notes

The implementation is conservative and maintains full compatibility with existing functionality. The key insight is that the `EngineRequest` struct already had the `JsonPayload` field, but the MutatingPolicy engine was not utilizing it. This PR bridges that gap by adding the necessary logic to detect and process JSON payloads while preserving all existing behavior for Kubernetes admission requests.

yrsuthari committed:
docs: improve API documentation clarity and fix typos

- Fix typos: 'v1apha1' and 'v1apha2' to 'v1alpha1' and 'v1alpha2'
- Add comprehensive documentation of all API groups used by Kyverno
- Clarify which versions belong to which API groups:
  - kyverno.io: v1beta1, v1, v2alpha1, v2beta1
  - policies.kyverno.io: v1alpha1
  - policyreport.io: v1alpha2
  - reports.kyverno.io: v1
- Update guidance for adding new resources to be API group specific

This improves developer experience by providing accurate and clear
information about Kyverno's API structure.

Signed-off-by: Yugandhar Suthari <yrsuthari@gmail.com>

yrsuthari committed:
feat: add JSON payload support to MutatingPolicy

- Modified Handle() and MatchedMutateExistingPolicies() methods in
  pkg/cel/policies/mpol/engine/engine.go to support JsonPayload field
- Added automatic detection of payload type (K8s admission vs JSON)
- Created minimal admission attributes for JSON payload processing
- Added comprehensive tests for JSON payload scenarios
- Added example documentation and usage examples

Fixes #13751

Signed-off-by: Yugandhar Suthari <yrsuthari@gmail.com>

yrsuthari committed:
feat: add JSON payload support to MutatingPolicy

- Enhanced Handle() and MatchedMutateExistingPolicies() methods in
  pkg/cel/policies/mpol/engine/engine.go to support JsonPayload field
- Added automatic detection of payload type (K8s admission vs JSON)
- Created minimal admission attributes for JSON payload processing
- Added comprehensive tests for JSON payload scenarios in engine_test.go
- Maintains full backward compatibility with existing functionality

Fixes #13751

Signed-off-by: Yugandhar Suthari <yrsuthari@gmail.com>

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

realshuting reviewed (commented):
Can you please a CLI tests?

realshuting commented on pkg/cel/policies/mpol/engine/engine.go:
```diff
@@ -180,33 +207,59 @@ func (e *engineImpl) handlePolicy(ctx context.Context, mpol Policy, attr admissi
 }
 
 func (e *engineImpl) MatchedMutateExistingPolicies(ctx context.Context, request engine.EngineRequest) []string {
```
Why do we care about whether it's a mutate existing policy or not? For JSON payloads, it will be evaluated via CLI and there's no concept of admission/existing resources.

realshuting reviewed (commented):
Please see comments.

yrsuthari committed:
Address review comments for JSON payload support in MutatingPolicy

- Fix MatchedMutateExistingPolicies to return nil for JSON payloads since there's no concept of 'mutate existing' for non-Kubernetes resources
- Add JSON payload support to CLI PolicyProcessor for MutatingPolicies
- Add comprehensive CLI tests for JSON payload mutations with both Kubernetes and non-Kubernetes JSON
- Improve test coverage with new JSON payload test cases

Addresses reviewer feedback on PR #13913

Signed-off-by: Yugandhar Suthari <yrsuthari@gmail.com>

yrsuthari committed:
Update generated client code and fix CLI test formatting

- Update generated clientset and informer files from codegen
- Fix CLI test file formatting for json-payload tests
- Ensure all generated code is up to date

Signed-off-by: Yugandhar Suthari <yrsuthari@gmail.com>

yrsuthari committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy

aerosouund committed:
Merge branch 'main' into feature/json-payload-support-mutating-policy
